src/main/java/network/brightspots/rcv/ContestConfigMigration.java

```diff
@@ -0,0 +1,216 @@
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
+ * the GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License along with this
+ * program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package network.brightspots.rcv;
+
+import static network.brightspots.rcv.Utils.isNullOrBlank;
+
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import network.brightspots.rcv.RawContestConfig.ContestRules;
+import network.brightspots.rcv.RawContestConfig.CvrSource;
+import network.brightspots.rcv.Tabulator.OvervoteRule;
+import network.brightspots.rcv.Tabulator.TieBreakMode;
+import network.brightspots.rcv.Tabulator.WinnerElectionMode;
+
+class ContestConfigMigration {
+
+  private static final Pattern versionNumPattern = Pattern.compile("(\\d+).*");
+
+  private static ArrayList<Integer> parseVersionString(String version) {
+    ArrayList<Integer> parsed = new ArrayList<>();
+    String[] arr = version.split("\\.");
+    for (String numString : arr) {
+      Matcher m = versionNumPattern.matcher(numString);
+      if (m.matches()) {
+        parsed.add(Integer.parseInt(m.group(1)));
+      }
+    }
+
+    return parsed;
+  }
+
+  // not intended to be used if either version is null
+  private static boolean isVersionNewer(String version1, String version2) {
+    if (version1.equals(version2)) {
+      return false;
+    }
+
+    ArrayList<Integer> version1Parsed = parseVersionString(version1);
+    ArrayList<Integer> version2Parsed = parseVersionString(version2);
+
+    for (int i = 0; i < version1Parsed.size(); i++) {
+      if (version2Parsed.size() <= i) {
+        return true;
+      }
+      int version1Num = version1Parsed.get(i);
+      int version2Num = version2Parsed.get(i);
+      if (version1Num > version2Num) {
+        return true;
+      } else if (version2Num > version1Num) {
+        return false;
+      }
+    }
+
+    return false;
+  }
+
+  static boolean isConfigVersionOlderThanAppVersion(String configVersion) {
+    return configVersion == null || isVersionNewer(Main.APP_VERSION, configVersion);
+  }
+
+  static boolean isConfigVersionNewerThanAppVersion(String configVersion) {
+    if (configVersion == null) {
+      return false;
+    }
+
+    if (isVersionNewer(configVersion, Main.APP_VERSION)) {
+      Logger.severe(
+          "Unable to process a config file with version %s using older version %s of the app!",
+          configVersion, Main.APP_VERSION);
+      return true;
+    }
+
+    return false;
+  }
+
+  static void migrateConfigVersion(ContestConfig config)
+      throws ConfigVersionIsNewerThanAppVersionException {
+    String version = config.rawConfig.tabulatorVersion;
+    if (version != null &&
+        (version.equals(Main.APP_VERSION) || version
+            .equals(ContestConfig.AUTOMATED_TEST_VERSION))) {
```

Comment:
I think I prefer it the way I have it now, because I was often finding it annoying that I'd open a test file and the app would want to migrate it. And it's easy enough to run bash commands to batch-update the test config files to streamline the workaround you described.